scio-tensorflow/src/main/scala/com/spotify/scio/tensorflow/TensorFlowDoFns.scala

```diff
@@ -111,16 +106,25 @@ sealed trait PredictDoFn[T, V, M <: Model[_]]
 
   @Teardown
   override def teardown(): Unit = {
-    Log.info(s"Tearing down predict DoFn $this")
-    val (running, m) = getResource().get(modelId)
-    if (running.decrementAndGet() == 0) {
-      m.close()
-    }
+    Log.info("Tearing down predict DoFn {}", this)
+    getResource.compute(
+      modelId,
+      {
+        case (_, null) =>
+          Log.warn("No model to close while tearing down predict DoFn")
+          null
+        case (_, (1, model)) =>
+          model.close()
+          null
+        case (_, (refCount, model)) =>
+          (refCount - 1) -> model
+      }
+    )
   }
 }
 
 object PredictDoFn {
-  type Resource[M <: Model[_]] = ConcurrentMap[String, (AtomicInteger, M)]
```

Comment:
`ConcurrentMap` does not ensure `compute` to be atomic. Replace with `ConcurrentHashMap` which does.

Reply:
Changing type here fails the binary compatibility check. Will revert to lower abstraction

Reply:
`override def createResource(): Resource[M] = new ConcurrentHashMap[String, (AtomicInteger, M)]()`

Won't the above code already take care of the right implementation to use?
I thought getResource will use resource created by `createResource`, of course except the AtomicInteger to Int change that has been introduced?

Reply:
That's correct. It would have worked because we make the **assumption** the implementation given by `createResource` has an atomic `compute`. Inspecting the resource javadoc on `compute` in `setup` and `teardown` tells this call is not guaranteed to be atomic. By changing the type, we explicitly require the resource to have an atomic `compute`.

In this case this is not really important because `PredictDoFn` is sealed and `createResource` is never overridden.

Reply:
Thanks for the reasoning. I really wondered why `ConcurrentMap` in a sealed trait. In a regular trait it would enable other concurrent maps like a Caffeine cache.